pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java

```diff
@@ -1661,54 +1669,144 @@ public static List<KitRequestShipping> getKitRequestsAfterBookmark(long bookmark
-                    if (StringUtils.isNotBlank(shortId) && shortId.length() > 5) {
-                        //Pepper HURID was used for upload
-                        id = shortId.trim();
-                    } else {
-                        //user was uploading with legacy id instead of new Pepper HURID
-                        //label creation job will call DDP go get information to generate collaborator_ids
-                        noGeneratedID = true;
-                    }
-                } else {
-                    //extract participant information from tissue collaborator_sample_id
-                    //RGP collaborator_sample_ids can be ignored! not in DSM
-                    if (collaboratorSampleId.contains("_")) {
-                        //get participant part from sample by looking for second "_"
-                        //not last "_" because user could have done something like PART_0001_Tissue_4
-                        collaboratorParticipantId =
-                                collaboratorSampleId.substring(0, collaboratorSampleId.indexOf("_", collaboratorSampleId.indexOf("_") + 1));
-                    }
-                }
-            }
-        } else {
-            if (StringUtils.isNotBlank(shortId)) {
-                id = shortId.trim();
-            } else {
-                //TODO if gen2 with connection check for shortId
-                id = ddpParticipantId.trim();
+        int instanceId = Integer.parseInt(latestKitRequest.getInstanceID());
+        DDPInstance ddpInstance = DDPInstance.getDDPInstanceById(instanceId);
+        return getCollaboratorParticipantId(ddpInstance, ddpParticipantId, shortId, collaboratorParticipantLengthOverwrite);
+    }
+
+    /**
+     * <p>
+     * Retrieves the collaborator participant ID for a given participant. The method first checks if the participant has kit with
+     * a legacy collaborator participant ID . If so that ID is returned. If the kit is a Gen2 migrated kit, a collaborator
+     * participant ID is generated based on previous Gen2 samples.
+     * Otherwise, a new collaborator participant ID is generated from Pepper short id.
+     *</p>
+     * <p>
+     * Called  during  kit creation through KitUpload, BSPDummyRoutes, KitLookUp
+     * and NonPepperKitCreationService
+     *      </p>
+     * @return the collaborator participant ID
+     */
+
+    public static String getCollaboratorParticipantId(DDPInstance ddpInstance, String ddpParticipantId, String shortId,
+                                                      String collaboratorParticipantLengthOverwrite) {
+        int instanceId = ddpInstance.getDdpInstanceIdAsInt();
+        String baseUrl = ddpInstance.getBaseUrl();
+        String collaboratorIdPrefix = ddpInstance.getCollaboratorIdPrefix();
+        //TODO PT collaboratorParticipantLengthOverwrite should be an int
+        if (ddpInstance.hasEsIndex()) {
```

Comment:
Why does the presence of an ES index correspond to appropriateness for handling the case of legacy participant ids?  I would have thought that just checking for legacy id would be sufficient.

Reply:
I understand your point, however having the migratedDDP set to true is not the best way to find this out, not only because I am not sure how early we started introducing and setting that field to true, but also OS2 does not have that value set to true but there might be some OS1 legacy participants that reconsented into OS2. 

Also to me as a developer, the only way to find the legacy short id is through ES, so for me this more means "if we have  a way to check for and get  access to the legacy ids"